nixos/lib/test-driver/test-driver.py

```diff
@@ -245,110 +229,252 @@ def _perform_ocr_on_screenshot(
-        self.shared_dir = create_dir("shared-xchg")
+    _cmd: str
 
-        self.booted = False
-        self.connected = False
-        self.pid: Optional[int] = None
-        self.socket = None
-        self.monitor: Optional[socket.socket] = None
-        self.allow_reboot = args.get("allowReboot", False)
+    def cmd(
+        self,
+        monitor_socket_path: pathlib.Path,
+        shell_socket_path: pathlib.Path,
+        allow_reboot: bool = False,  # TODO: unused, legacy?
+    ) -> str:
+        display_opts = ""
+        display_available = any(x in os.environ for x in ["DISPLAY", "WAYLAND_DISPLAY"])
+        if not display_available:
+            display_opts += " -nographic"
+
+        # qemu options
+        qemu_opts = ""
+        qemu_opts += (
+            ""
+            if allow_reboot
+            else " -no-reboot"
+            " -device virtio-serial"
+            " -device virtconsole,chardev=shell"
+            " -serial stdio"
+        )
+        # TODO: qemu script already catpures this env variable, legacy?
+        qemu_opts += " " + os.environ.get("QEMU_OPTS", "")
+
+        return (
+            f"{self._cmd}"
+            f" -monitor unix:{monitor_socket_path}"
+            f" -chardev socket,id=shell,path={shell_socket_path}"
+            f"{display_opts}"
+        )
 
     @staticmethod
-    def create_startcommand(args: Dict[str, str]) -> str:
-        net_backend = "-netdev user,id=net0"
-        net_frontend = "-device virtio-net-pci,netdev=net0"
+    def build_environment(
+        state_dir: pathlib.Path,
+        shared_dir: pathlib.Path,
+    ) -> os._Environ:
+        env = os.environ
+        env.update(
+            {
+                "TMPDIR": str(state_dir),
+                "SHARED_DIR": str(shared_dir),
+                "USE_TMPDIR": "1",
+            }
+        )
+        return env
+
+    def run(
+        self,
+        state_dir: pathlib.Path,
+        shared_dir: pathlib.Path,
+        monitor_socket_path: pathlib.Path,
+        shell_socket_path: pathlib.Path,
+    ) -> subprocess.Popen:
+        return subprocess.Popen(
+            self.cmd(monitor_socket_path, shell_socket_path),
+            stdin=subprocess.DEVNULL,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            shell=True,
+            cwd=state_dir,
+            env=self.build_environment(state_dir, shared_dir),
+        )
 
-        if "netBackendArgs" in args:
-            net_backend += "," + args["netBackendArgs"]
 
-        if "netFrontendArgs" in args:
-            net_frontend += "," + args["netFrontendArgs"]
+class NixStartScript(StartCommand):
+    """A start script from nixos/modules/virtualiation/qemu-vm.nix
+    that also satisfies the requirement of the BaseStartCommand.
+    These nix command have the particular charactersitic that the
+    machine name can be extracted out of them via a regex match.
+    (Admittedly a _very_ implicit contract, evtl. TODO fix)
```

Comment:
@roberth I'd like to get your eyes back on deck here, since this is our angle of approach (for not to say "attack" :wink:) to solve the additional restriction on hostnames over RFC dns lables.

We can't fix that without fixing the way that the hostnames are recovered from within the test driver. (What about FQDN's with clashing host labels?)

Once fixed, this entire class should go away.

Reply:
If I remember correctly; on the NixOS side, all we need is [system.name](https://search.nixos.org/options?channel=unstable&show=system.name&from=0&size=50&sort=relevance&query=system.name) to be our python identifier and `networking.hostName` to do its usual thing.

However, we need to get both names. I don't think the interface between python and Nix is set up for that, judging by the awkward parsing. Do we generate a nice json on the Nix side yet?

Reply:
We could change the `startScripts` argument type from `[Path]` to `[(MachineName, Path)]` where `MachineName` is a `String`. The real machine name could then be injected at buildtime by Nix to the Python driver. The Python driver could then sanitize this machine name to create a valid Python variable name (`. -> _`, `pass -> pass_`, ... ). (This would also allow us to create a nice helping message to indicate the machine name transformation that has been applied.) 
Moreover, the real machine name could still be used as keys in the `machines` dict.

I could address this in a followup PR.

Reply:
Let's avoid full-on "sanitization" though. It'd be something that the caller has to learn and something we may have to break in the future. If we require the caller to use python identifiers in the `machines` attrset, they're still free to set `network.hostName`.
We can default the `network.hostName` to the python identifier, with `_` swapped for `-`. If you consider that sanitization, then touché, but it is well delineated because python identifiers are quite restrictive. Hostnames, in contrast, can have both `.` and `-`, making the reverse translation non-injective. That would make me sad :(. At least we can take this simplified view, because we're only talking about a default value. Restrictions on the first character make the picture more complicated, which is why it's great to defer the final naming decisions to the user, who already has to deal with restrictions anyway. And those are the restrictions that are inherent to _their_ test, not the restrictions we try to predict.

Reply:
My bad, i actually forgot the machine has a name in the node attribute set! So, i agree we don't need to sanitize it since the caller could have to choose a name which is also a valid Python identifier.

However, instead of parsing the outpath, i think we could still pass the machine name as script arguments.

Reply:
I think, we should probably starting on resolving this issue immediately after getting this PR through since it promises another cleanup of the code path.

Maybe indeed the best here is to use the machine name of the `nodes` attribute as exported symbols, but also hold fqdn in a `machines` dict.

Reply:
:+1: on scoping this out.

It'd be nice to have a new NixOS option `system.labels = mkOption {type = json;}` for the "sole" purpose of being written to a file `$toplevel/labels.json`. A nice interface for passing data from `config` to NixOS 'consumers'. The name borrows from k8s, fwiw. Not `system.`_`nixos`_`.labels` because `nix-darwin` should be able to do the same.